### PR TITLE
fix(factories): CreateShorthandOptions should be typed

### PR DESCRIPTION
**Pull request by lucivpav**
Warning: ugly code ahead

### CODE REVIEW

**Comment by lucivpav** [comment]
@layershifter please let me know what you think

File: docs/src/prototypes/MenuButton/MenuButton.tsx [hunk 1]
```diff
@@ -12,6 +12,7 @@ import {
   Alignment,
   Position,
   UNSTABLE_Popper,
+  WithExpectedProps,
```
Comment (layershifter) [inline docs/src/prototypes/MenuButton/MenuButton.tsx]: Let's remove this thing, there is no real need in it.

File: packages/react/src/lib/factories.ts [hunk 2]
```diff
@@ -8,18 +8,19 @@ import {
...
 import { mergeStyles } from './mergeThemes'
 
 type HTMLTag = 'iframe' | 'img' | 'input'
 type ShorthandProp = 'children' | 'src' | 'type'
 
-interface CreateShorthandOptions {
+interface CreateShorthandOptions<P> {
   /** Default props object */
-  defaultProps?: Props
+  defaultProps?: P
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  defaultProps?: Partial<Props<P>>
```

- `Partial`: We don't need to specify all default props
- `Props<P>`: We should be able to specify additional props

File: packages/react/src/lib/factories.ts [hunk 3]
```diff
@@ -8,18 +8,19 @@ import {
...
 
   /** Override props object or function (called with regular props) */
-  overrideProps?: Props & ((props: Props) => Props) | Props
+  overrideProps?: P & ((props: P) => P) | P
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  overrideProps?: Partial<Props<P>> | ((props: P) => Partial<Props<P>>)
```

https://github.com/stardust-ui/react/pull/1886/files#r322256487

File: packages/react/src/lib/factories.ts [hunk 4]
```diff
@@ -28,12 +29,6 @@ interface CreateShorthandOptions {
   render?: ShorthandRenderFunction
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  render?: ShorthandRenderFunction<P>
```

File: packages/react/src/lib/factories.ts [hunk 5]
```diff
@@ -28,12 +29,6 @@ interface CreateShorthandOptions {
...
 }
 
-const CREATE_SHORTHAND_DEFAULT_OPTIONS: CreateShorthandOptions = {
-  defaultProps: {},
-  overrideProps: {},
-  generateKey: true,
-}
-
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts:36]: 👍

File: packages/react/src/lib/factories.ts [hunk 6]
```diff
@@ -45,21 +40,34 @@ const mappedProps: { [key in HTMLTag]: ShorthandProp } = {
...
+  children?
+  className?
+  style?
+  styles?
+  key?
+  kind?
+  onChange?
+  onFocus?
+  onMouseDown?
+}
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Let's remove it

File: packages/react/src/lib/factories.ts [hunk 7]
```diff
@@ -45,21 +40,34 @@ const mappedProps: { [key in HTMLTag]: ShorthandProp } = {
...
-  options = CREATE_SHORTHAND_DEFAULT_OPTIONS,
+  options,
 }: {
   Component: React.ElementType
   allowsJSX?: boolean
   mappedProp?: string
   mappedArrayProp?: string
-  valueOrRenderCallback?: ShorthandValue<Props> | ShorthandRenderCallback
-  options?: CreateShorthandOptions
+  valueOrRenderCallback?: ShorthandValue<P> | ShorthandRenderCallback
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  valueOrRenderCallback?: ShorthandValue<P> | ShorthandRenderCallback<P>
```

File: packages/react/src/lib/factories.ts [hunk 8]
```diff
@@ -45,21 +40,34 @@ const mappedProps: { [key in HTMLTag]: ShorthandProp } = {
...
+  options?: CreateShorthandOptions<WithExpectedProps<P>>
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  options?: CreateShorthandOptions<P>
```

File: packages/react/src/lib/factories.ts [hunk 9]
```diff
@@ -93,7 +101,7 @@ type CreateShorthandFactoryConfigInner<TPropName = string> = {
 export type CreateShorthandFactoryConfig = CreateShorthandFactoryConfigInner
 export type ShorthandFactory<P> = (
   value: ShorthandValue<P>,
-  options?: CreateShorthandOptions,
+  options?: CreateShorthandOptions<WithExpectedProps<P>>,
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  options?: CreateShorthandOptions<P>,
```

File: packages/react/src/lib/factories.ts [hunk 10]
```diff
@@ -162,15 +170,15 @@ function createShorthandFromValue({
   mappedProp?: string
   mappedArrayProp?: string
   allowsJSX?: boolean
-  value?: ShorthandValue<Props>
-  options?: CreateShorthandOptions
+  value?: ShorthandValue<P>
+  options?: CreateShorthandOptions<WithExpectedProps<P>>
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  options?: CreateShorthandOptions<P>
```

File: packages/react/src/lib/factories.ts [hunk 11]
```diff
@@ -162,15 +170,15 @@ function createShorthandFromValue({
...
 }) {
   if (typeof Component !== 'function' && typeof Component !== 'string') {
     throw new Error('createShorthand() Component must be a string or function.')
   }
   // short circuit noop values
   const valIsNoop = _.isNil(value) || typeof value === 'boolean'
-  if (valIsNoop && !options.render) return null
+  if (valIsNoop && options && !options.render) return null
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  if (valIsNoop && options.render) return null
```

File: packages/react/src/lib/factories.ts [hunk 12]
```diff
@@ -45,21 +40,34 @@ const mappedProps: { [key in HTMLTag]: ShorthandProp } = {
...
 /** A more robust React.createElement. It can create elements from primitive values. */
-export function createShorthand({
+export function createShorthand<P>({
   allowsJSX,
   Component,
   mappedProp,
   mappedArrayProp,
   valueOrRenderCallback,
-  options = CREATE_SHORTHAND_DEFAULT_OPTIONS,
+  options,
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  options = {},
```

File: packages/react/src/lib/factories.ts [hunk 13]
```diff
@@ -207,22 +215,25 @@ function createShorthandFromValue({
   // ----------------------------------------
   // Build up props
   // ----------------------------------------
-  const { defaultProps = {} } = options
+  const defaultProps = options ? options.defaultProps : undefined
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  const defaultProps = options.defaultProps || {} as Props<P>
```

File: packages/react/src/lib/factories.ts [hunk 14]
```diff
@@ -207,22 +215,25 @@ function createShorthandFromValue({
...
   // User's props
   const usersProps = (valIsReactElement && {}) || (valIsPropsObject && (value as Props)) || {}
 
   // Override props
-  let { overrideProps } = options
+  let overrideProps = options ? options.overrideProps : undefined
   overrideProps =
     typeof overrideProps === 'function'
-      ? (overrideProps as Function)({ ...defaultProps, ...usersProps })
+      ? ((overrideProps as unknown) as Function)({ ...defaultProps, ...usersProps })
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```tsx
  const overrideProps: Props<P> =
    typeof options.overrideProps === 'function'
      ? options.overrideProps({ ...defaultProps, ...usersProps }) as Props<P>
      : options.overrideProps as Props<P> || {} as Props<P>
```

File: packages/react/src/lib/factories.ts [hunk 15]
```diff
@@ -207,22 +215,25 @@ function createShorthandFromValue({
...
       : overrideProps || {}
 
   // Merge props
   const props = { ...defaultProps, ...usersProps, ...overrideProps }
 
-  const mappedHTMLProps = mappedProps[overrideProps.as || defaultProps.as]
+  const mappedHTMLProps =
+    mappedProps[
+      (overrideProps as WithAsProp<P>).as || (defaultProps && (defaultProps as WithAsProp<P>).as)
+    ]
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 16]
```diff
@@ -207,22 +215,25 @@ function createShorthandFromValue({
...
-  let { overrideProps } = options
+  let overrideProps = options ? options.overrideProps : undefined
   overrideProps =
     typeof overrideProps === 'function'
-      ? (overrideProps as Function)({ ...defaultProps, ...usersProps })
+      ? ((overrideProps as unknown) as Function)({ ...defaultProps, ...usersProps })
       : overrideProps || {}
 
   // Merge props
   const props = { ...defaultProps, ...usersProps, ...overrideProps }
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  const props: Props<P> = { ...defaultProps, ...usersProps, ...overrideProps }
```

File: packages/react/src/lib/factories.ts [hunk 17]
```diff
@@ -235,29 +246,37 @@ function createShorthandFromValue({
   }
 
   // Merge className
-  if (defaultProps.className || overrideProps.className || usersProps.className) {
+  if ((defaultProps && defaultProps.className) || overrideProps.className || usersProps.className) {
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 18]
```diff
@@ -235,29 +246,37 @@ function createShorthandFromValue({
...
     const mergedClassesNames = cx(
-      defaultProps.className,
+      defaultProps && defaultProps.className,
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 19]
```diff
@@ -235,29 +246,37 @@ function createShorthandFromValue({
...
 
   // Merge style
-  if (defaultProps.style || overrideProps.style || usersProps.style) {
-    props.style = { ...defaultProps.style, ...usersProps.style, ...overrideProps.style }
+  if ((defaultProps && defaultProps.style) || overrideProps.style || usersProps.style) {
+    props.style = {
+      ...(defaultProps && defaultProps.style),
+      ...usersProps.style,
+      ...overrideProps.style,
+    }
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 20]
```diff
@@ -235,29 +246,37 @@ function createShorthandFromValue({
...
 
   // Merge styles
-  if (defaultProps.styles || overrideProps.styles || usersProps.styles) {
-    props.styles = mergeStyles(defaultProps.styles, usersProps.styles, overrideProps.styles)
+  if ((defaultProps && defaultProps.styles) || overrideProps.styles || usersProps.styles) {
+    props.styles = mergeStyles(
+      defaultProps && defaultProps.styles,
+      usersProps.styles,
+      overrideProps.styles,
+    )
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 21]
```diff
@@ -235,29 +246,37 @@ function createShorthandFromValue({
...
   }
 
   // ----------------------------------------
   // Get key
   // ----------------------------------------
-  const { generateKey = true } = options
+  const generateKey = options && options.generateKey !== undefined ? options.generateKey : true
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 22]
```diff
@@ -285,7 +304,7 @@ function createShorthandFromValue({
   // ----------------------------------------
   // Create Element
   // ----------------------------------------
-  const { render } = options
+  const render = options ? options.render : undefined
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]: Can be reverted

File: packages/react/src/lib/factories.ts [hunk 23]
```diff
@@ -315,9 +334,9 @@ function createShorthandFromRenderCallback({
   mappedProp?: string
   mappedArrayProp?: string
   allowsJSX?: boolean
-  options?: CreateShorthandOptions
+  options?: CreateShorthandOptions<WithExpectedProps<P>>
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
  options?: CreateShorthandOptions<P>
```

File: packages/react/src/lib/factories.ts [hunk 24]
```diff
@@ -93,7 +101,7 @@ type CreateShorthandFactoryConfigInner<TPropName = string> = {
 export type CreateShorthandFactoryConfig = CreateShorthandFactoryConfigInner
```
Comment (layershifter) [inline packages/react/src/lib/factories.ts]:
```suggestion
```

Comment (layershifter) [reply packages/react/src/lib/factories.ts]: `CreateShorthandFactoryConfigInner` can be also deleted

File: packages/react/src/lib/factories.ts (hunk 16 above)
Comment (lucivpav) [reply packages/react/src/lib/factories.ts]: P doesn't have `className`, `style`, `styles`, `key` though

Comment (layershifter) [reply packages/react/src/lib/factories.ts]: It's `Props<P>`, `Props<P> = P & any` 🤔

Comment (lucivpav) [reply packages/react/src/lib/factories.ts]: Sorry I meant `Props<P>`. I cannot do `props.className` if `const props: Props<P>`

File: docs/src/prototypes/meetingOptions/components/MSTeamsLogo.tsx [hunk 25]
```diff
@@ -18,6 +18,7 @@ class MSTeamsLogo extends React.Component<LogoProps, any> {
             <div style={style}>
               {Icon.create(icon, {
                 defaultProps: {
+                  name: 'teams',
```
Comment (layershifter) [inline docs/src/prototypes/meetingOptions/components/MSTeamsLogo.tsx]: It's not more required and can be reverted

File: packages/react/src/components/Accordion/Accordion.tsx [hunk 26]
```diff
@@ -265,7 +265,6 @@ class Accordion extends AutoControlledComponent<WithAsProp<AccordionProps>, Acco
             index,
             contentRef,
             canBeCollapsed,
-            id: titleId,
```
Comment (layershifter) [inline packages/react/src/components/Accordion/Accordion.tsx]: Oops, I think that is intentional. Can we revert this change?

File: packages/react/src/components/Accordion/AccordionTitle.tsx [hunk 27]
```diff
@@ -120,6 +120,7 @@ class AccordionTitle extends UIComponent<WithAsProp<AccordionTitleProps>, any> {
           {...applyAccessibilityKeyHandlers(accessibility.keyHandlers.content, unhandledProps)}
           start={Icon.create(indicatorWithDefaults, {
             defaultProps: {
+              name: 'stardust-circle',
```
Comment (layershifter) [inline packages/react/src/components/Accordion/AccordionTitle.tsx]: Same thing about all `Icon.create()`, we don't need to specify it under `defaultProps`

File: packages/react/src/components/Accordion/Accordion.tsx (hunk 26 above)
Comment (lucivpav) [reply packages/react/src/components/Accordion/Accordion.tsx]: Same goes for line 280?

File: packages/react/src/components/Toolbar/ToolbarMenuItem.tsx [hunk 28]
```diff
@@ -108,7 +108,9 @@ class ToolbarMenuItem extends UIComponent<WithAsProp<ToolbarMenuItemProps>> {
           children
         ) : (
           <>
-            {Icon.create(icon, { defaultProps: { xSpacing: !!content ? 'after' : 'none' } })}
+            {Icon.create(icon, {
+              defaultProps: { name: 'stardust-circle', xSpacing: !!content ? 'after' : 'none' },
```
Comment (layershifter) [inline packages/react/src/components/Toolbar/ToolbarMenuItem.tsx]: Should be also reverted

**Review by layershifter — APPROVED** [review]
LGTM 👍
